software/py/bsg_manycore_link_gen.py

```diff
@@ -0,0 +1,252 @@
+
+      if re.search(".dram$", sec) != None:
+        continue
+
+      if re.search(".dmem$", sec):
+        in_objects = '*'
+      else:
+        sec += '.dmem'
+      
+      if sec != ".dmem":
+        prev_sec = section_map[i-1][0]
+
+        if re.search(".dmem$", prev_sec) == None:
+          prev_sec += '.dmem'
+
+        laddr = "LOADADDR({0}) + ADDR({1}) - ADDR({0})".format(
+            prev_sec, sec)
+
+      sections += self._section(sec, laddr, 'DMEM_VMA', None,
+          in_sections, in_objects)
+
+    # DRAM sections
+    for i,m in enumerate(section_map):
+      sec = m[0]
+      vaddr = ""
+      in_sections = m[1]
+      in_objects = '*'
+
+      if re.search(".dmem$", sec) != None:
+        continue
+
+      if sec == ".text.dram":
+        vaddr = "0x0"
+
+      if re.search(".dram$", sec) == None and self._default_data_loc == 'dram':
+        sec += '.dram'
+
+      if self._default_data_loc == 'dram' or re.search(".dram$", sec) != None:
+        sections += self._section(sec, vaddr, None, 'DRAM_LMA',
+            in_sections, in_objects)
+
+      if sec == '.text.dram':
+        sections += ". = . + 0x80000000;\n\n"
+
+    # Symbols
+    if self._default_data_loc == 'dmem':
+      sections += "_gp = ADDR(.sdata.dmem) + 0x800;\n"
+    else:
+      sections += "_gp = ADDR(.sdata.dram) + 0x800;\n"
+    sections += "_sp = 0x{0:08x};\n".format(self._sp)
+    sections += "_bsg_data_start_addr = 0x{0:08x};\n".format(_DMEM_VMA_START)
+    sections += "_bsg_data_end_addr = ADDR(.striped.data.dmem) + " \
+                "SIZEOF(.striped.data.dmem);\n"
+    sections += "_bsg_striped_data_start = ADDR(.striped.data.dmem)\n;"
+    sections += "_bsg_dram_t_start_addr = LOADADDR(.text.dram)\n;"
+    sections += "_bsg_dram_d_start_addr = LOADADDR(.text.dram) + SIZEOF(.text.dram)\n;"
+    sections += "_bsg_dram_d_end_addr = LOADADDR(.dram) + SIZEOF(.dram)\n;"
+    sections += "_bsg_dram_end_addr = _bsg_dram_d_end_addr\n;"
+    sections += "_end = _bsg_dram_end_addr\n;"
+
+    sections += "\n}\n"
+
+    script = self._opening_comment + '\n'
+    script += self._memory_regions(mem_regions) + '\n'
+    script += sections
+    return script
+
+
+if __name__ == '__main__':
+  # Parse arguments
+  parser = argparse.ArgumentParser(prog = 'bsg_manycore_link_gen.py',
+      formatter_class = argparse.RawDescriptionHelpFormatter,
+      description = bsg_manycore_link_gen.__doc__)
+  parser.add_argument('--default_data_loc', 
+    help = 'Default data location',
+    default = 'dmem',
+    choices = ['dmem', 'dram'])
+  parser.add_argument('--dram_size', 
+    help = 'DRAM size',
+    default = 0x80000000,
+    type = int)
+  parser.add_argument('--sp', 
+    help = 'Stack pointer',
+    default = 0x1000,
+    type = lambda x: int(x, 0))
+  args = parser.parse_args()
+
+
+  # Generate linker script
+  link_gen = bsg_manycore_link_gen(args.default_data_loc, args.dram_size,
+      args.sp)
+  print(link_gen.script())
```

Comment:
I think it would be useful to be able to specify the output file as an argument. If it isn't provided, you can print it to stdout.

Reply:
Done.